onnxruntime/core/providers/cpu/tensor/gather_elements.cc

```diff
@@ -0,0 +1,276 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+#include "gather_elements.h"
+
+namespace onnxruntime {
+
+ONNX_CPU_OPERATOR_KERNEL(
+    GatherElements,
+    11,
+    KernelDefBuilder().TypeConstraint("T", DataTypeImpl::AllTensorTypes())
+                      .TypeConstraint("Tind", std::vector<MLDataType>{DataTypeImpl::GetTensorType<int32_t>(), 
+                                                                      DataTypeImpl::GetTensorType<int64_t>()}),
+    GatherElements);
+
+// Some helpers needed for GatherElements op -
+
+// The following method computes the offset in the flattened array
+// using every axis except the inner dimension (as the offset is just 1)
+// and the axis that 'GatherElements' is processing for as that requires the corresponding
+// 'indices' value
+// This prevents the need to compute this offset for every element within the same 'inner_dimension' chunk
+// as this value just differs by 1 for the chunk elements and we can have this cached and re-use as needed
+int64_t compute_base_offset(const std::vector<int64_t>& shape, const TensorPitches& pitches, int64_t skip_axis) {
+  // in this context, rank can never be < 1, so saving checking overhead
+  auto loop_size = static_cast<int64_t>(shape.size()) - 1;
+
+  int64_t base_offset = 0;
+
+  for (int64_t i = 0; i < loop_size; ++i) {
+    if (i != skip_axis)
+      base_offset += (shape[i] * pitches[i]);
+  }
+
+  return base_offset;
+}
+
+// This method computes the number of 'inner_dimension' chunks
+// Example: input = [2, 3]     output = 2
+//          input = [3, 2, 4]  output = 3 * 2 = 6
+//          input  = [2]       output = 1
+int64_t calculate_num_inner_dim(const std::vector<int64_t>& dims) {
+  // in this context, rank can never be < 1, so saving checking overhead
+  int64_t num_inner_dims = 1;
+
+  for (int64_t i = 0; i < static_cast<int64_t>(dims.size()) - 1; ++i) {
+    num_inner_dims *= dims[i];
+  }
+
+  return num_inner_dims;
+}
+
+// This method computes increments over an 'inner_dimension'
+// Example 1: current_dims = [0, 0] tensor_dims = [3, 1], then current_dims = [1, 0]
```

Comment:
Example 1: It looks like tensor_dims = [3, 1] should be populated in all lines. [2, 1] is incorrect. Let me know if that makes things clearer.

Example 2: The first step should be clear. We are "skipping over" a block of inner_dim to make it [0, 1, 0]. In the next step (increment), it becomes [0, 2, 0] (but then the shape is actually [1, 2, x]), so we would have to reset the middle dim and increment the outermost dim making it [1, 0, 0], but then again since the shape is [1, 2, x], we would have to reset the outermost dim to 0 as well.

Reply:
Adding to the previous comment - If you think about it, it makes sense. The tensor dims  = [1, 2, 2] meaning there are 2 inner_dims (each of size 2). We can "skip over" one inner_dim. But when we try to "skip over" the other inner_dim and land on the first element after it - there are no elements beyond that and so we "roll back" to the starting point. This is the "intuition" behind this.